akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpEntity.scala

```diff
@@ -507,15 +507,43 @@ object HttpEntity {
       withSizeLimit(SizeLimit.Disabled)
 
     override def transformDataBytes(transformer: Flow[ByteString, ByteString, Any]): HttpEntity.Chunked = {
-      val newData =
-        chunks.map {
-          case Chunk(data, "")    => data
-          case LastChunk("", Nil) => ByteString.empty
-          case _ =>
-            throw new IllegalArgumentException("Chunked.transformDataBytes not allowed for chunks with metadata")
-        } via transformer
-
-      HttpEntity.Chunked.fromData(contentType, newData)
+      import akka.stream.scaladsl.GraphDSL.Implicits._
+
+      val buf = scala.collection.mutable.ArrayBuffer.empty[Throwable]
+
+      val transformChunks: Flow[ChunkStreamPart, ChunkStreamPart, NotUsed] = Flow.fromGraph(GraphDSL.create() { implicit builder: GraphDSL.Builder[NotUsed] =>
+
+        val bcast = builder.add(Broadcast[ChunkStreamPart](2))
+        val concat = builder.add(Concat[ChunkStreamPart](2))
+
+        val newDataF: Flow[ChunkStreamPart, ByteString, Any] = Flow[ChunkStreamPart].map {
+          case Chunk(data, "")  => data
+          case LastChunk("", _) => ByteString.empty
+        }
+          .via(transformer).recover {
+            case t: Throwable =>
+              buf += t
+              ByteString.empty
+          }
+
+        val toChunksF: Flow[ByteString, ChunkStreamPart, Any] = Flow[ByteString].collect {
+          case b: ByteString if b.nonEmpty => Chunk(b)
+        }
+
+        val trailHeaderF: Flow[ChunkStreamPart, ChunkStreamPart, Any] = Flow[ChunkStreamPart].collect {
+          case lc @ LastChunk(_, s) if s.nonEmpty => lc
+        }.recover {
+          case _: Throwable =>
+            // just needs to complete
+            LastChunk
+        }
+
+        bcast ~> newDataF ~> toChunksF ~> concat
+        bcast ~> trailHeaderF ~> concat
+        FlowShape(bcast.in, concat.out)
+      })
+
+      HttpEntity.Chunked(contentType, chunks.via(transformChunks).map(c => if (buf.nonEmpty) throw buf.head else c))
```

Comment:
Ah, I see, errors should be thrown here. That's a somewhat uncommon approach that might not be thread-safe depending on how streams are executed.

Reply:
Well LastChunk doesn't carry any data so its not the subject of processing with the `transformer` Flow. 

`bcast ~> trailHeaderF ~> concat` will fail only when theres bad encoding (for example: akka/http/scaladsl/server/directives/CodingDirectivesSpec.scala:117) but we don't care because in such case `bcast ~> newDataF ~> toChunksF ~> concat` will fail as well with same exception.

What we care about is the potential failure of `.via(transformer).` and that is just being written to the buffer.

Maybe you can elaborate more on the thread safety issue because I don't see it (yet :smile:  )